atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1138 @@
+    ///     Ok(()) => println!("Metrics published successfully"),
+    ///     Err(e) => println!("Failed to publish metrics: {}", e),
+    /// }
+    /// ```
+    ///
+    /// # Security Considerations
+    ///
+    /// - Metrics are signed to ensure authenticity and prevent tampering
+    /// - Uses CBOR for efficient binary serialization
+    /// - Published through gossipsub's secure message propagation
+    /// - The hash includes all critical metrics data to ensure integrity
+    #[instrument(
+        level = "info",
+        fields(
+            node_small_id = %usage_metrics.node_small_id,
+            node_public_url = %usage_metrics.node_public_url,
+        ),
+        skip_all
+    )]
+    fn handle_new_usage_metrics_event(
+        &mut self,
+        usage_metrics: NodeUsageMetrics,
+    ) -> Result<(), AtomaP2pNodeError> {
+        let NodeUsageMetrics {
+            node_public_url,
+            node_small_id,
+            country,
+            timestamp,
+        } = usage_metrics;
+        let usage_metrics_hash = blake3::hash(
+            &[
+                node_public_url.as_bytes(),
+                &node_small_id.to_le_bytes(),
+                country.as_bytes(),
+                &timestamp.to_le_bytes(),
+            ]
+            .concat(),
+        );
+        let active_sui_address = self.keystore.addresses()[0];
+        let signature = self
+            .keystore
+            .sign_hashed(&active_sui_address, usage_metrics_hash.as_bytes())
+            .map_err(|e| {
+                error!(
+                    target = "atoma-p2p",
+                    event = "sign_usage_metrics_hash_error",
+                    error = %e,
+                    "Failed to sign usage metrics hash"
+                );
+                AtomaP2pNodeError::SignatureError(e.to_string())
+            })?;
+        let usage_metrics = UsageMetrics {
+            node_public_url,
+            node_small_id,
+            country,
+            timestamp,
+            signature: signature.as_ref().to_vec(),
+        };
+        let mut serialized_message = Vec::new();
+        ciborium::into_writer(&usage_metrics, &mut serialized_message).map_err(|e| {
+            error!(
+                target = "atoma-p2p",
+                event = "serialize_usage_metrics_error",
+                error = %e,
+                "Failed to serialize usage metrics"
+            );
+            AtomaP2pNodeError::UsageMetricsSerializeError(e)
+        })?;
+        let topic = gossipsub::IdentTopic::new(METRICS_GOSPUBSUB_TOPIC);
+        self.swarm
+            .behaviour_mut()
+            .gossipsub
+            .publish(topic, serialized_message)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum AtomaP2pNodeError {
+    #[error("Failed to build gossipsub config: {0}")]
+    GossipsubConfigError(#[from] ConfigBuilderError),
+    #[error("Failed to build gossipsub: {0}")]
+    GossipsubBuildError(String),
```

Comment:
These two errors seems to address the same thing

Reply:
Addressed